mtp_transaction_uploader/upload.py

```diff
@@ -313,26 +313,48 @@ def get_matching_batch_id_for_settlement(record):
         # not a worldpay settlement
         return
 
+    relative_date = record.date.date()
+    batch_date = m.group('date')
     try:
-        batch_date = datetime.datetime.strptime(m.group(1), '%d%m').date()
+        if len(batch_date) == 4:
+            batch_date = parse_4_digit_date(batch_date, relative_date)
+        elif len(batch_date) == 2:
+            batch_date = parse_2_digit_date(batch_date, relative_date)
+        else:
+            # no date provided so cannot match to a batch
+            raise ValueError
     except ValueError:
         # settlement date cannot be parsed
         return
 
-    relative_date = record.date.date()
-
-    # parse 4-digit date
-    batch_date = batch_date.replace(year=relative_date.year)
-    if batch_date > relative_date:
-        batch_date = batch_date.replace(year=relative_date.year - 1)
-
     # get batch id for date if found
     conn = get_authenticated_connection()
     response = conn.batches.get(date=batch_date.isoformat())
     if response.get('results'):
         return response['results'][0]['id']
 
 
+def parse_2_digit_date(date_str, relative_date: datetime.date) -> datetime.date:
+    batch_date = datetime.datetime.strptime(date_str, '%d').date()
+    batch_date = batch_date.replace(year=relative_date.year, month=relative_date.month)
+    if batch_date <= relative_date:
```

Comment:
Probably not a huge difference in practice but I'm wondering whether we should loose the `=` year.
AFAICT the batch date shouldn't really be the same as the `relative_date`/day on the statement - this date is always at least 1-2 days in the past.

I don't imagine making much of a difference/causing problems but I thought I'd mention anyway just-in-case (similar situation in `parse_4_digit_date`)

Reply:
are you suggesting doing
```
batch_date = batch_date.replace(month=relative_date.month)
```
instead of
```
batch_date = batch_date.replace(year=relative_date.year, month=relative_date.month)
```

i'm not sure that's a good idea. `batch_date` would automatically get today's year, but `relative_date` might actually not be from this year. i think in practice, it doesn't matter hugely because we always read the bank statement the day after it's created. the purpose of having `relative_date` date at all was so that the transaction's date is used from the bank statement to calculate the settlement date as opposed to "today". it allows for the system to read old statements (if ever necessary) and prevents funny business around new year.

Reply:
> are you suggesting doing

```python
batch_date = batch_date.replace(month=relative_date.month)
```

🤔 I was wondering whether we should change this 

```python
if batch_date <= relative_date:
  return ...
```

into

```python
if batch_date < relative_date:
  return ...
```

so `<=` => `<` to make it more strict, e.g. if today/relative date is the 22nd March and the string is `22` it would be interpreted as 22nd February.

Reply:
oh i misunderstood what you were pointing out.

are you suggesting `if batch_date < relative_date:` instead of `if batch_date <= relative_date:`? i kept the logic the same as it was before. i.e. it allows for a settlement date to be the same date as that of the transaction.

https://github.com/ministryofjustice/money-to-prisoners-transaction-uploader/blob/a2167c34ece6a2eff3183eed0b383689da777821/mtp_transaction_uploader/upload.py#L316
… this only decremented the year if the statement date seemed to be _strictly_ in the future

Reply:
i do see the point that a settlement is unlikely to be posted on exactly the date it's for… but _could_ it somehow happen? i'm slightly uncertain about changing this. i don't know why initially it was allowed so i thought it was best to leave it